### ISSUE TITLE
No real coediting

### ISSUE THREAD

**Issue by Sicoval31** [nextcloud/documentserver_community#302]
Nextcloud 25.0.7 (snap auto upgrade)
office nextcloud 7.8.0
Community document server 0.1.13

since few months, when editing a file (docx, xlsx) with OO and many partners, the collaborate editing mode doesn't work fine :
Partner A add info in a Excel cell, partner B show wich cell, but never see what is added, even if Partner A save the file.

Need to quit OO and reload the file.
try with Firefox, edge or chrome .. same problem
try with other computer : same problem

No error message

It's not a collaborate mode

Thanks to repair or explain why coediting is not allow

**Comment by zeptomoon** [comment]
This seems to be a duplicate of #288 . 
[QUOTE:] _"Cursor of co-editors is displayed but changes aren't visible"_

**Comment by mattiscb** [comment]
Nextcloud 27.0.2,
Community document server 0.1.13

We face the same issue since the update to 27.0.2. It worked fine in NC 26 we had before. Users must quit and enter OnlyOffice to see work of others.

And yes #288 refers to the same problem, but this issuer also refers here.

**Comment by Sicoval31** [comment]
Any news to solve this problem ?

**Comment by chadek** [comment]
Did you try the last documentserver_community version released few weeks ago?

**Comment by Sicoval31** [comment]
Yes, installing the last 0.1.15 version. No change

**Comment by chadek** [comment]
Fixed in v0.1.16

**Comment by Sicoval31** [comment]
Don't work. Same problem
Why closing directly whitout users testing ????

**Comment by chadek** [comment]
Right, sorry, I was trying to close duplicates of [this issue](https://github.com/nextcloud/documentserver_community/issues/100) that v0.1.16 aim to fix and this is clearly an other issue.

**Comment by chadek** [comment]
It's been here for so long, it should have been fix in latest build. Would be nice if someone can confirm and test it more deeply before closing it.

**Comment by chadek** [comment]
Co-authoring works now, and **v0.4.0** is where I would say it is actually fixed rather than mostly working.

**What was wrong.** Two separate things, and your description — "partner B sees which cell, but never sees what is added" — is precisely the first one. Cursor and selection messages were reaching the other browser; the *changes* were not being displayed. The broadcast this server sends when someone saves a batch of changes was missing the chunk flags sdkjs uses to decide whether a batch is complete, so the receiving editor buffered every batch forever and rendered none of it. Everything looked right on the wire. That was fixed in **v0.3.2/v0.3.3**.

What v0.4.0 adds is the reliability the first fix exposed:

* **a second participant no longer takes the document away from everyone.** sdkjs sends a `close` command from `DocsCoApi.disconnect()` whenever an editor drops to view mode with the page still open — a licence verdict on the number of users in a document, rights being taken away, `disconnectEveryone`, a critical error — and this server was treating that as "the editor was closed": it removed the session and, when it was the last one, wrote the document out and deleted its folder, underneath a page that was still editing. Everything that session did afterwards then replayed against a baseline that had been thrown away. The command is advisory now.
* **a batch of changes that collides on `change_index` is retried.** Two co-authors saving at the same moment aimed for the same indexes; the loser got an integrity-constraint violation thrown out of a transaction nobody closed, which left the connection in a dead transaction, so *every later query in that request* failed with `SAVEPOINT DOCTRINE_n does not exist` — the session's saves kept failing after the collision, not just the one that raced. Measured on the rig with four processes writing to one document: before, 38 of 120 changes stored and 41 of 60 batches failed; after, 120 of 120, no duplicates, no gaps, no failures.
* **the memcache IPC backend no longer loses an in-flight message.** A reader could claim a slot whose payload had not been written yet, return nothing, and make that message unreachable for good while everything queued behind it moved up a place. That one only bites instances with `memcache.distributed` set, which is most of the bigger ones.

**Proof.** `tests/rig/coedit_test.py` drives two headless browsers, logs a different user into each, opens the same file in both, types in each, and then asks *each editor's own document model* what it displays — not whether the messages were sent. Against `master` just now, Nextcloud 34, connector 10.1, bundled Euro-Office 9.3.4:

```
===== LIVE PROPAGATION =====
   A sees: '... FLUA2 \r\n A1 \r\n TABB2 ...'
   A: all 4 markers present
   B sees: '... FLUA2 \r\n A1 \r\n TABB2 ...'
   B: all 4 markers present
   VERDICT: live co-authoring OK

>>> session A (admin): 1 socket session(s) established -> ['f3836787d018719a']  (more than one means it reconnected mid-edit)
>>> session B (bob): 1 socket session(s) established -> ['2f605963220164cf']  (more than one means it reconnected mid-edit)
-- coedit: PASS in 197s
```

And the same in a spreadsheet, which is the case you both reported — changes there carry recalculation indexes and used to be the worse of the two:

```
   A sees: 'U61=XA1  | U67=XB1 '
   A: all 2 markers present
   B sees: 'U61=XA1  | U67=XB1 '
   B: all 2 markers present
   VERDICT: live co-authoring OK
-- coedit-xlsx: PASS in 140s
```

`tests/rig/twotab_test.py` covers the second-participant case, including one of them being dropped to view mode:

```
==> round 2
   A co-auth: {"isCoAuthoring":true,"countEditUsers":2,"bufferedChunks":0,"bufferedOther":0,"changesIndex":318,...}
   B co-auth: {"isCoAuthoring":true,"countEditUsers":2,"bufferedChunks":0,"bufferedOther":0,"changesIndex":318,...}
   tab A is missing: nothing
   tab B is missing: nothing

==> tab B is dropped to view mode (what sdkjs does on a licence verdict, rights change or disconnectEveryone)
   doc folders: 1 | changes: 318 | sessions: 2
   A types on, and it must still reach the file
   file: {'TAB49343-A9': True}

   VERDICT: two tabs co-author
-- twotab: PASS in 269s
```

`bufferedChunks: 0` and `bufferedOther: 0` are the numbers that were non-zero and growing when this was broken: that is sdkjs holding changes it had received and would not render.

Both tests run in CI on pull requests and weekly, so this does not quietly regress again.
